pycloudlib/azure/cloud.py

```diff
@@ -0,0 +1,644 @@
+        self.username = "ubuntu"
+
+        self.registered_instances = {}
+        self.registered_images = {}
+
+        config_dict = {}
+
+        if client_id:
+            config_dict["clientId"] = client_id
+
+        if client_secret:
+            config_dict["clientSecret"] = client_secret
+
+        if subscription_id:
+            config_dict["subscriptionId"] = subscription_id
+
+        if tenant_id:
+            config_dict["tenantId"] = tenant_id
+
+        self.resource_client = util.get_client(
+            ResourceManagementClient, config_dict
+        )
+
+        self.network_client = util.get_client(
+            NetworkManagementClient, config_dict
+        )
+
+        self.compute_client = util.get_client(
+            ComputeManagementClient, config_dict
+        )
+
+        self.resource_group = self._create_resource_group()
+        self.base_tag = tag
+
+    def _create_network_security_group(self):
+        """Create a network security group.
+
+        This method creates a network security groups that allows the user
+        to ssh into the machine and execute commands.
+
+        Returns:
+            The network security object created by Azure
+
+        """
+        security_group_name = "{}-sgn".format(self.tag)
+        nsg_group = self.network_client.network_security_groups
+
+        self._log.debug('Creating Azure network security group')
+        nsg_call = nsg_group.create_or_update(
+            resource_group_name=self.resource_group.name,
+            network_security_group_name=security_group_name,
+            parameters={
+                "location": self.location,
+                "security_rules": [
+                    {
+                        "name": "SSH",
+                        "properties": {
+                            "priority": 300,
+                            "protocol": "TCP",
+                            "access": "Allow",
+                            "direction": "Inbound",
+                            "sourceAddressPrefix": "*",
+                            "sourcePortRange": "*",
+                            "destinationAddressPrefix": "*",
+                            "destinationPortRange": "22"
+                        }
+                    }
+                ]
+            }
+        )
+
+        return nsg_call.result()
+
+    def _create_resource_group(self):
+        """Create a resource group.
+
+        This method creates an Azure resource group. Every other componet that
+        we create will be contained into this resource group. This means that
+        if we delete this resource group, we delete all resources associated
+        with it.
+
+        Returns:
+            The resource group created by Azure
+
+        """
+        resource_name = "{}-rg".format(self.tag)
+        self._log.debug('Creating Azure resource group')
+
+        return self.resource_client.resource_groups.create_or_update(
+            resource_name,
+            {
```

Comment:
Can we provide tags here too so each resource we create is tagged the same (it aids in finding and removing separate resources in the future)